middleware.go

```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
+	o := evaluateMiddlewareOptions(opts)
+	return loggingInterceptor{logger: logger, middlewareOptions: o}.interceptStream
+}
+
+type loggingInterceptor struct {
+	logger *logrus.Logger
+	*middlewareOptions
+}
+
+type GRPCRequest struct {
+	Method    string `json:"method,omitempty"`
+	UserAgent string `json:"userAgent,omitempty"`
+	PeerAddr  string `json:"peer,omitempty"`
+	Deadline  string `json:"deadline,omitempty"`
+	Duration  string `json:"duration,omitempty"`
+}
+
+func (l loggingInterceptor) intercept(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (resp interface{}, err error) {
+	startTime := time.Now()
+	ctx = ctxlogrus.ToContext(ctx, logrus.NewEntry(l.logger))
+
+	request := l.requestFromContext(ctx, &GRPCRequest{
+		Method: info.FullMethod,
+	})
+
+	resp, err = handler(ctx, req)
+
+	request.Duration = fmt.Sprintf("%.9fs", time.Since(startTime).Seconds())
+
+	l.log(ctx, err, info.FullMethod)
+
+	return
+}
+
+func (l loggingInterceptor) interceptStream(
+	srv interface{},
+	ss grpc.ServerStream,
+	info *grpc.StreamServerInfo,
+	handler grpc.StreamHandler,
+) (err error) {
+	startTime := time.Now()
+	ctx := ctxlogrus.ToContext(ss.Context(), logrus.NewEntry(l.logger))
+
+	request := l.requestFromContext(ctx, &GRPCRequest{
+		Method: info.FullMethod,
+	})
+
+	wrapped := grpc_middleware.WrapServerStream(ss)
+	wrapped.WrappedContext = ctx
+
+	err = handler(srv, wrapped)
+
+	request.Duration = fmt.Sprintf("%.9fs", time.Since(startTime).Seconds())
+
+	l.log(ctx, err, info.FullMethod)
+
+	return
+}
+
+func (l *loggingInterceptor) requestFromContext(ctx context.Context, request *GRPCRequest) *GRPCRequest {
+	if d, ok := ctx.Deadline(); ok {
+		request.Deadline = d.UTC().Format(time.RFC3339Nano)
+	}
 
-			log.WithFields(fields).Info("Completed request")
+	if p, ok := peer.FromContext(ctx); ok && p != nil {
+		u := &url.URL{
+			Scheme: p.Addr.Network(),
+			Host:   p.Addr.String(),
+		}
+		request.PeerAddr = u.String()
+	}
+
+	if md, ok := metadata.FromIncomingContext(ctx); ok && md != nil {
+		request.UserAgent = strings.Join(md.Get("user-agent"), "")
+
+		if trace := md.Get("X-Cloud-Trace-Context"); len(trace) == 1 {
+			ctxlogrus.AddFields(ctx, logrus.Fields{"trace": trace[0]})
+		}
+	}
+
+	ctxlogrus.AddFields(ctx, logrus.Fields{"grpcRequest": request})
+
+	return request
+}
+
+// logStatus adds the gRPC Status to the log context.
+// If the response is an internal server error, log that as an Error
+// returns true if the logging was handled (e.g. internal server error)
+func (l *loggingInterceptor) log(ctx context.Context, err error, method string) {
+	if !l.filterRPC(ctx, method, err) {
+		return
+	}
+
+	if err != nil {
+		st := status.Convert(err)
+		jsonStatus, merr := ProtoJSONMarshalOptions.Marshal(st.Proto())
+		if merr != nil {
+			ctxlogrus.Extract(ctx).WithError(merr).Errorf("error marshalling error status into log")
```

Comment:
should we return here?

Also, let's only log something at error if we think it indicates an important bug that deserves immediate attention. This is just logging an error about logging

```suggestion
			ctxlogrus.Extract(ctx).WithError(merr).Warnf("error marshalling error status into log")
```

Reply:
Yep good call, fixing that